File: Node/src/dto/enums/AppointmentType.ts
import { UUID } from "../tsPrimitives/UUID";
import { ValidateNested } from "class-validator";

export type AppointmentType = AppointmentType.Complex   | 
                              AppointmentType.Procedure

export namespace AppointmentType {

    export class Procedure {
        @ValidateNested()
        procedure: UUID

        constructor(id: UUID) {
            this.procedure = id
        }
    }

    export class Complex {
        @ValidateNested()
        complex: UUID

        constructor(id: UUID) {
            this.complex = id
        }
    }
}

File: android-legacy/shared/src/main/kotlin/shared/dto/enums/AppointmentType.kt
@file:UseSerializers(UUIDSerializer::class)

package shared.dto.enums

import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import shared.dto.objects.procedure.ChunkId
import shared.dto.protocols.Responsable
import shared.serializers.AppointmentTypeSerializer
import shared.serializers.UUIDSerializer
import java.util.*

typealias ProcedureId = UUID

@Serializable(AppointmentTypeSerializer::class)
sealed class AppointmentType : Responsable {
    @Serializable
    data class Complex(
        val id: UUID,
        val order: Map<ChunkId, ProcedureId>
    ) : AppointmentType()

    @Serializable
    data class Procedure(
        val id: UUID,
    ) : AppointmentType()
}